src/lib/unlang/xlat_eval.c

```diff
@@ -222,6 +222,121 @@ static inline void xlat_debug_log_result(request_t *request, fr_value_box_t cons
 	RDEBUG2("  --> %pV", result);
 }
 
+/** Process list of boxed values provided as input to an xlat
+ *
+ * Ensures that the value boxes passed to an xlat function match the
+ * requirements listed in its "args", and escapes any tainted boxes
+ * using the specified escaping routine.
+ *
+ * @param[in] ctx		in which to allocate any buffers.
+ * @param[in,out] in		value boxes provided as input.
+ * @param[in] input_type	required by xlat.
+ * @param[in] args		definition of arguments required by xlat.
+ */
+static xlat_action_t xlat_process_args(TALLOC_CTX *ctx, fr_value_box_list_t *in, request_t *request,
+					xlat_input_type_t input_type, xlat_arg_parser_t args[])
+{
+	xlat_arg_parser_t	*arg;
+	fr_value_box_t		*vb, *next;
+
+	/*
+	 *	No args registered for this xlat
+	 */
+	if (!args) return XLAT_ACTION_DONE;
+
+	switch (input_type) {
+	case XLAT_INPUT_UNPROCESSED:
+		/*
+		 *	xlat needs no input processing just return.
+		 */
+		return XLAT_ACTION_DONE;
+
+	case XLAT_INPUT_MONO:
+		/*
+		 *	xlat takes all input as a single vb.
+		 */
+		vb = fr_dlist_head(in);
+		if (vb) {
+			if (args->concat) {
+				/* Concat all vbs to one, casting to the required type. */
+				if (fr_value_box_list_concat(ctx, vb, in, args->type, true) < 0) {
+concat_error:
+					RERROR("Failed concatenating input");
+					return XLAT_ACTION_FAIL;
+				}
+			} else {
+				if ((vb->type == args->type) && (vb->type != FR_TYPE_VALUE_BOX)) return XLAT_ACTION_DONE;
+
+				/* Just cast the first vb to the required type */
+				if (fr_value_box_cast_in_place(ctx, vb, args->type, NULL) < 0) {
+cast_error:
+					RERROR("Failed converting input to %s",
+						fr_table_str_by_value(fr_value_box_type_table, args->type, "<INVALID>"));
+					return XLAT_ACTION_FAIL;
+				}
+			}
+		} else {
+			if (args->required) {
+missing:
+				RERROR("Missing required argument");
+				return XLAT_ACTION_FAIL;
+			}
+		}
+
+		break;
+
+	case XLAT_INPUT_ARGS:
+		/*
+		 *	xlat consumes a sequence of arguments.
+		 */
+		arg = args;
+		vb = fr_dlist_head(in);
+		while (arg->type != FR_TYPE_INVALID) {
+			/*
+			 *	pre-advance, in case the vb is replaced during processing
+			 */
+			next = fr_dlist_next(in, vb);
+
+			if ((arg->required) && (!vb)) goto missing;
+
+			if (!vb) break;
+
+			// *todo* make this escape tainted value boxes
+			if ((arg->concat) && (vb->type == FR_TYPE_GROUP)) {
+				fr_value_box_t *new_vb = fr_value_box_alloc(ctx, arg->type, NULL, false);
+				if (fr_value_box_list_concat(ctx, new_vb, &vb->vb_group, arg->type, true) < 0) {
+					talloc_free(new_vb);
+					goto concat_error;
+				}
+				fr_dlist_replace(in, vb, new_vb);
+				talloc_free(vb);
+			} else if ((vb->type == FR_TYPE_GROUP) && (arg->type != FR_TYPE_GROUP)) {
```

Comment:
I don't think moving up a level is correct here as it'd make it impossible for the xlat to determine the position of arguments after this one. Most xlats that take multiple boxes for an argument want to iterate over them as multiple values for that particular argument.  Casting in place is definitely correct though.

Reply:
There should probably be a validation pass when registering argument arrays which checks that the type passed in is in the set `FR_TYPE_VALUE`.